### ISSUE TITLE
FR: Async requests

### ISSUE THREAD

**Issue by eranhirs** [firebase/firebase-admin-python#104]
I am using `firebase-admin` on an `asyncio` project and when there are a lot of requests it slows down the whole API.

Surprisingly, I didn't see any other issues about this, I hope I am not duplicating.

```python
ref = yield from db.reference
result = yield from ref.get()
```

Thanks

**Comment by hiranya911** [comment]
Can this be not handled in the application code? Basically execute Admin SDK methods like `ref.get()` on separate threads.

**Comment by eranhirs** [comment]
It won't scale as I would like and, to my opinion, complicates the code and is harder to track.

**Comment by hiranya911** [comment]
Most of our users are on Python 2.7, so we cannot directly expose async coroutines from the SDK anytime soon. In the meantime you should be able to get around the scalability issues by using the [`run_in_executor()`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.run_in_executor) method of the asyncio event loop.

```
ref = db.reference(path) # Doesn't make RPC calls

executor = concurrent.futures.ThreadPoolExecutor(max_workers=8)
loop = asyncio.get_event_loop()
loop.run_in_executor(executor, ref.get) # Make RPC call on a separate thread
```

**Comment by eranhirs** [comment]
Thanks! I will consider using it.

Meanwhile, I believe this should be reopened and prioritized.
[There are more users](https://stackoverflow.com/questions/46378139/aiohttp-and-firebase-and-firebase-admin-is-it-non-blocking-python-3-6) who need this.
It is a standard in python 3. It's not like python 2.7 that there are a lot of different libraries to support.

How about supporting both by exposing different methods?

**Comment by eranhirs** [comment]
In case anyone runs into the same problem -
Eventually, I am using for now the [REST API](https://firebase.google.com/docs/database/rest/retrieve-data).
Calling it asynchronously with `aiohttp`.

**Comment by Natim** [comment]
@eranhirs would you feel like starting a async-firebase project that we could contribute to?

**Comment by hyperknot** [comment]
I wanted to report that having async support for an auth SDK is very important. This is exactly the area where async frameworks make a huge difference, where you do a client side request inside a server side request.

Please consider adding core support and examples for using with ASGI frameworks, like Starlette:
https://github.com/encode/starlette

**Comment by Lawouach** [comment]
I agree with the general feeling here. Not having async support is making this library a tad less useful nowadays. Not very future proof unfortunately.

**Comment by labroid** [comment]
+1 on this. My project has a few hundred thousand items that need to be worked via Firebase, and ~1 second latency times make wait times intractable without async support

**Comment by labroid** [comment]
Python 2.7 is no longer supported. Any progress here?

**Comment by hiranya911** [comment]
This needs to be first addressed in other GCP Python libraries that we use as dependencies. More specifically, the google-auth library (which we use to make all authorized HTTP calls) needs to support it. From what I can see, they have plans to add async support starting from 2020, but there is no concrete timetable as of now:

https://github.com/googleapis/google-auth-library-python/issues/321
https://github.com/googleapis/google-cloud-python/issues/3103

I can only imagine that this will take a while considering the engineering effort that is involved. Pretty much all existing APIs have to be re-evaluated and redesigned for async support. 

If google-auth takes a very long time to introduce async support, we can consider wrapping the existing google-auth APIs in async threads, so that at least firebase-admin can expose an async API to the developer. But even that is going to require a lot of work, based on some prototyping results I've managed to put together.

**Comment by kamyar** [comment]
Is there any development on this front? 
Also can you please reopen this issue?
We have been using firestore and lack of async is a very big downside of firestore.
If async supported is not planned for google services, It would be nice if we know beforehand so we can plan our migration to services that support it. Lack of this feature is causing us both money and product quality tbh.

**Comment by hiranya911** [comment]
There's been no development in this space afaik. We cannot support an async API until the core GCP Python libraries like `google-auth` provide async support. For Firestore you can file a feature request directly at https://github.com/googleapis/python-firestore

**Comment by kamyar** [comment]
It seems there is an [open PR](https://github.com/googleapis/python-api-core/pull/29#pullrequestreview-423782507) related to adding async support.

One of the contributors to the libraries just left [a comment](https://github.com/googleapis/google-cloud-python/issues/3103#issuecomment-638355699). 

Would be great to hear about firebase-admin maintainers/contributors perspective on this. :)

**Comment by hiranya911** [comment]
Good to know that `python-api-core` is in the process of adding async support. After that we need at least [`google-auth`](https://github.com/googleapis/google-auth-library-python) to adopt asyncio before we can provide async support (we'd very much like to do so btw). I see that you're already following their issue at https://github.com/googleapis/google-auth-library-python/issues/321.

Firestore APIs are entirely developed in the `python-firestore` repo I linked above, and it's their call to support asyncio or not. But I think having async support `python-api-core` makes it feasible for them to support it.

**Comment by kamyar** [comment]
Great to hear these, thanks a lot @hiranya911 :) 
I will look around in the `python-firestore` for the discussion, and maybe check if they have something planned. 🚀

**Comment by bradreardon** [comment]
It seems that an async-based auth flow was added to `google-auth` in September of last year: https://github.com/googleapis/google-auth-library-python/blob/master/CHANGELOG.md#1220-2020-09-28
https://github.com/googleapis/google-auth-library-python/pull/612

Are there any plans to make this library async-compatible now that its dependencies seem to have done so?

**Comment by lahirumaramba** [comment]
Good news! We are prioritizing this task and have plans to start the initial planning work over the coming months. It is still too early to provide a timeline, but I will reopen this issue so we can track the progress here. Thank you everyone for your patience!

**Comment by miksa** [comment]
Hi! Is there any update about this feature ? I saw a pull request to use new version of the google-auth api, would that fulfill prerequisites ?

**Comment by guneemwelloeux** [comment]
Hello! I know the last comment wasn't left here long ago, but as it didn't get any answer, I'm posting another one.
Has a timeline been agreed upon since April 2021, and if there is one, could you please let everyone know here?

Many thanks!

**Comment by lahirumaramba** [comment]
Thank you again for your patience on this!
This is a high priority task for us and we should start seeing the initial changes sometime in Q2/Q3 2022. I can't promise you a timeline for this feature but rest assured that this is something on our immediate roadmap for this year and you will see updates in the coming months.

**Comment by jkyle109** [comment]
We just released [v5.3.0](https://github.com/firebase/firebase-admin-python/releases/tag/v5.3.0) which added async support for the Cloud Firestore API. This is the first step into fulfilling this FR so please give it a try and let us know what you think!

**Comment by antont** [comment]
Seems to be working well, we haven't used this under heavy load yet, though.

I'm curious to hear how others are experiencing this, but AFAIK it's good and robust.

From my pov this is completed, we mostly use firestore and do async gcloud storage uploads too, but I don't know if something is missing still otherwise.

**Comment by alexandermalyga** [comment]
Hi, are there any plans to add async support to the admin auth client? Thanks for the great work!

**Comment by Natim** [comment]
If you want async support, you should rely on google-cloud-firestore:

```
from google.cloud.firestore_v1.async_client import AsyncClient

db = AsyncClient()

async def add_doc_async():
    doc_ref = db.collection("users").document()
    await doc_ref.set({"name": "Alice"})
```

**Comment by hojjatabdollahi** [comment]
@Natim Thank you. But does that work for auth too?

**Comment by lahirumaramba** [comment]
Firebase Admin Python SDK supports [Firestore Async](https://firebase.google.com/support/release-notes/admin/python?#cloud-firestore_2) clients. See https://firebase.google.com/docs/firestore/quickstart#python_2

The SDK also supports [async counterparts](https://firebase.google.com/support/release-notes/admin/python?#cloud-messaging_1) for Messaging (`send_each_async()` and `send_each_for_multicast_async()`).

We will continue to extend support for other services in the future. If you have requests for async support in a specific Firebase service please create a new FR covering that service. It helps us track progress instead of having a global FR for all the services/APIs.
